Source/Details/ASCollectionLayoutSpecCalculator.h

```diff
@@ -0,0 +1,47 @@
+//
+//  ASCollectionLayoutSpecCalculator.h
+//  AsyncDisplayKit
+//
+//  Created by Huy Nguyen on 2/3/17.
+//  Copyright © 2017 Facebook. All rights reserved.
+//
+
+#import <AsyncDisplayKit/ASCollectionLayoutCalculating.h>
+#import <AsyncDisplayKit/ASDimension.h>
+#import <AsyncDisplayKit/ASScrollDirection.h>
+
+@class ASLayoutSpec;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface ASCollectionLayoutSpecCalculator : NSObject <ASCollectionLayoutCalculating>
+
+/// Scrollable directions of the containing collection view. Default to vertical directions.
+@property (nonatomic, assign, readonly) ASScrollDirection scrollableDirections;
+
+- (instancetype)initWithScrollableDirections:(ASScrollDirection)scrollableDirections NS_DESIGNATED_INITIALIZER;
+
+/**
+ * @abstract Return a size range, derived from the given size and scroll directions, that will be used to calculate the collection view layout.
+ *
+ * @param viewportSize The viewport size that the resulting layout should fit, usually equals to the size of the containing collection view
+ */
+- (ASSizeRange)sizeRangeForLayoutSpecThatFits:(CGSize)viewportSize;
+
+/**
+ * @abstract Return a layout spec that describes the collection view layout with regard to the given elements.
+ * The returned layout spec is used to calculate an ASLayout.
+ *
+ * @param sizeRange The size range that the resulting layout should fit. 
+ * This is the size range returned by -sizeRangeForLayoutSpecThatFits:
+ *
+ * @param map An element map containings all elements to be laid out.
+ *
+ * @discussion This method can be called on multiple theads and thus must be thread-safe (and ideally, stateless).
+ * This method must block the calling thread but can dispatch to other theads to reduce the blocking time.
+ */
+- (ASLayoutSpec *)layoutSpecThatFits:(ASSizeRange)sizeRange forElementMap:(ASElementMap *)map;
```

Comment:
I think we can eliminate this class and put the burden of doing this on the client, probably. For instance, in the flow layout case, it would amount to:

```objc
- (ASLayout *)layoutWithSize:(CGSize)size elementMap:(ASElementMap *)map
{
   ASLayoutSpec *spec = [ASStackLayoutSpec …];
   // I know I am a vertical layout, so unbounded height: 
   ASSizeRange sizeRange = ASSizeRangeMake(size.width, 0, size.width, INFINITY);
   return [spec layoutThatFits:sizeRange];
}
```

- There's not much extra code.
- Gives the calculator class control over scrollable directions.
- Eliminates a class.